Unable to refresh iceberg table due to NPE

<!-- (At least include the following, feel free to add if you have more content) -->

### Steps to reproduce the behavior (Required)
1. Run a refresh on an iceberg table
2. the logs will show an NPE

### Expected behavior (Required)
- Refreshing table metadata works

### Real behavior (Required)
It throws the following error: 
```
2025-11-04 13:17:31.167-05:00 WARN (com.starrocks.connector.hive.ConnectorTableMetadataProcessor|37) [CachingIcebergCatalog.refreshCatalog():418] refresh iceberg_express.some_table metadata cache failed, msg :
 com.google.common.util.concurrent.UncheckedExecutionException: java.lang.NullPointerException
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2085)
	at com.google.common.cache.LocalCache.get(LocalCache.java:4011)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:4034)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:5010)
	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:5017)
	at com.starrocks.connector.iceberg.CachingIcebergCatalog.refreshTable(CachingIcebergCatalog.java:377)
	at com.starrocks.connector.iceberg.CachingIcebergCatalog.refreshTable(CachingIcebergCatalog.java:353)
	at com.starrocks.connector.iceberg.CachingIcebergCatalog.refreshCatalog(CachingIcebergCatalog.java:415)
	at com.starrocks.connector.hive.ConnectorTableMetadataProcessor.refreshIcebergCachingCatalog(ConnectorTableMetadataProcessor.java:170)
	at com.starrocks.connector.hive.ConnectorTableMetadataProcessor.runAfterCatalogReady(ConnectorTableMetadataProcessor.java:116)
	at com.starrocks.common.util.FrontendDaemon.runOneCycle(FrontendDaemon.java:78)
	at com.starrocks.common.util.Daemon.run(Daemon.java:98)
Caused by: java.lang.NullPointerException
```


Update: I was able to get a better stack trace: 

```
Caused by: java.lang.NullPointerException: Cannot invoke "org.apache.iceberg.PartitionSpec.fields()" because "partitionSpec" is null
	at com.starrocks.connector.PartitionUtil.convertIcebergPartitionToPartitionName(PartitionUtil.java:783)
	at com.starrocks.connector.iceberg.IcebergCatalog.getPartitions(IcebergCatalog.java:332)
	at com.starrocks.connector.iceberg.CachingIcebergCatalog.lambda$new$2(CachingIcebergCatalog.java:121)
	at com.google.common.cache.CacheLoader$FunctionToCacheLoader.load(CacheLoader.java:169)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3570)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2312)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2189)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2079)
```


### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
`4.0.0-dccb665`

--- a/fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergCatalog.java
+++ b/fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergCatalog.java
@@ -348,6 +348,11 @@ public interface IcebergCatalog extends MemoryTrackable {
                             int specId = row.get(1, Integer.class);
                             PartitionSpec spec = nativeTable.specs().get(specId);
 
+                            // Old partition specs may be referenced in metadata even if they have been deleted. Skip them.
+                            if (spec == null) {
+                                continue;
+                            }
+
                             String partitionName =
                                     PartitionUtil.convertIcebergPartitionToPartitionName(nativeTable, spec, partitionData);
                             long lastUpdated =

--- a/fe/fe-core/src/test/java/com/starrocks/connector/iceberg/IcebergCatalogTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/connector/iceberg/IcebergCatalogTest.java
@@ -1,0 +1,81 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.connector.iceberg;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.starrocks.catalog.IcebergTable;
+import mockit.Mock;
+import mockit.MockUp;
+import org.apache.iceberg.DataTask;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.MetadataTableType;
+import org.apache.iceberg.MetadataTableUtils;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.PartitionsTable;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableScan;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.util.StructProjection;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+import java.util.Map;
+
+public class IcebergCatalogTest {
+    @Test
+    public void testGetPartitionsSkipsMissingSpecId() {
+        IcebergCatalog catalog = Mockito.mock(IcebergCatalog.class, Mockito.CALLS_REAL_METHODS);
+        Table nativeTable = Mockito.mock(Table.class);
+        PartitionSpec currentSpec = Mockito.mock(PartitionSpec.class);
+        Mockito.when(currentSpec.isUnpartitioned()).thenReturn(false);
+        Mockito.when(nativeTable.spec()).thenReturn(currentSpec);
+        Mockito.when(nativeTable.specs()).thenReturn(ImmutableMap.of(1, currentSpec));
+        Mockito.when(nativeTable.name()).thenReturn("db.tbl");
+
+        IcebergTable icebergTable = new IcebergTable(1, "srTable", "iceberg_catalog",
+                "resource", "db", "tbl", "", Lists.newArrayList(), nativeTable, Maps.newHashMap());
+
+        PartitionsTable partitionsTable = Mockito.mock(PartitionsTable.class);
+        TableScan scan = Mockito.mock(TableScan.class);
+        Mockito.when(partitionsTable.newScan()).thenReturn(scan);
+
+        FileScanTask task = Mockito.mock(FileScanTask.class);
+        DataTask dataTask = Mockito.mock(DataTask.class);
+        StructLike row = Mockito.mock(StructLike.class);
+        StructProjection partitionData = Mockito.mock(StructProjection.class);
+        Mockito.when(task.asDataTask()).thenReturn(dataTask);
+        Mockito.when(row.get(0, StructProjection.class)).thenReturn(partitionData);
+        Mockito.when(row.get(1, Integer.class)).thenReturn(99);
+        CloseableIterable<StructLike> rowIterable = CloseableIterable.withNoopClose(Lists.newArrayList(row));
+        Mockito.when(dataTask.rows()).thenReturn(rowIterable);
+
+        CloseableIterable<FileScanTask> taskIterable = CloseableIterable.withNoopClose(Lists.newArrayList(task));
+        Mockito.when(scan.planFiles()).thenReturn(taskIterable);
+
+        new MockUp<MetadataTableUtils>() {
+            @Mock
+            public Table createMetadataTableInstance(Table table, MetadataTableType type) {
+                return partitionsTable;
+            }
+        };
+
+        Map<String, Partition> partitions = catalog.getPartitions(icebergTable, -1, null);
+        Assertions.assertTrue(partitions.isEmpty());
+    }
+}

